client/gutenberg/extensions/simple-payments/edit.js

```diff
@@ -40,18 +40,44 @@ class SimplePaymentsEdit extends Component {
 		isSavingProduct: false,
 	};
 
+	/**
+	 * We'll use this flag to inject attributes one time when the product entity is loaded.
+	 *
+	 * It is based on the presence of a `paymentId` attribute.
+	 *
+	 * If present, initially we are waiting for attributes to be injected.
+	 * If absent, we may save the product in the future but do not need to inject attributes based
+	 * on the response as they will have come from our product submission.
+	 */
+	shouldInjectPaymentAttributes = !! this.props.attributes.paymentId;
+
 	componentDidMount() {
+		// Try to get the simplePayment loaded into attributes if possible.
 		this.injectPaymentAttributes();
+
+		const { attributes, hasPublishAction } = this.props;
+		const { paymentId } = attributes;
+
+		// If the user can publish save an empty product so that we have an ID and can save
+		// concurrently with the post that contains the Simple Payment.
+		if ( ! paymentId && hasPublishAction ) {
+			this.saveProduct();
+		}
 	}
 
 	componentDidUpdate( prevProps ) {
 		const { hasPublishAction, isSelected } = this.props;
 
-		if ( prevProps.simplePayment !== this.props.simplePayment ) {
+		if ( ! isEqual( prevProps.simplePayment, this.props.simplePayment ) ) {
 			this.injectPaymentAttributes();
 		}
 
-		if ( ! prevProps.isSaving && this.props.isSaving && hasPublishAction ) {
+		if (
+			! prevProps.isSaving &&
+			this.props.isSaving &&
+			hasPublishAction &&
+			this.validateAttributes()
```

Comment:
This is a bit tricky…

Now with draft products, if they're modified but still not valid then we'll never update the draft and any changes (stored in attributes) will be overwritten when the saved draft product is read from the API. Try adding an invalid product, updating it, then refreshing the post editor.

We might:

- Need to know what status the product is in
- Update drafts with invalid data
- Update published products only if they're valid